avocado/core/test.py

```diff
@@ -321,7 +321,8 @@ class Test(unittest.TestCase, TestData):
     timeout = None
 
     def __init__(self, methodName='test', name=None, params=None,
-                 base_logdir=None, job=None, runner_queue=None, tags=None):
+                 base_logdir=None, job=None, runner_queue=None, tags=None,
+                 requirements=None):
```

Comment:
Passing `tags` to a test was based on a requirement to use tags as "default" values in some situations (the arch on a qemu test, for instance).  I think it's better to keep the test unaware of the requirements, because the whole point is for the test to not even be started if the requirements are missing, and to be able to count on them by default.

Also, given that this is mostly a nrunner related feature, adding this information to the `Avocado.Test` class and to the (older) `avocado.core.loader` code may not be a good idea wrt the support we'll have to add, and then eventually drop.

Reply:
@willianrampazzo I thought more about this, and while many use test cases may not need to reference its requirements (for instance, it may need a package `foo` installed, but it will really just trust it's present and execute `foo-cmd`) there are some situations such as having a file downloaded (like the current assets) in which the test will need to reference them.

I don't have the right solution for that, but maybe something like and `id` in the requirements file:

```python
{'url': 'https://host/file.rom', 'id': 'rom'}
```

And on the test:

```python
class Asset(Test):
    def test(self):
        boot_with_rom(os.environ.get(`AVOCADO_REQUIREMENT__ROM`)
```

The point of using an environment variable, rather than adding support for the test is based on:

1) `exec-test` and similars (`tap`) won't have another way of accessing those requirements
2) current `INSTRUMENTED` tests are being kept compatible, but we may want/need to break compatibility when the current runner is replaced.

@beraldoleal do you have comments here?

Reply:
From my point of view, the _requirements resolver_ will be responsible for collecting/fulfilling the requirements for a test. This action, as you previously mentioned, will allow a test to start, depending on the availability of all the requirements. Initially, thinking from a high level, I would not expect the _requirements resolver_ to provide any information to the test (the opposite later in this comment [1]).

Looking at the test execution, how the test will use those requirements is the responsibility of the test, and not the _requirements resolver_. What I mean is, if an asset was downloaded before the test starts by the _requirements resolver_, it is available in the cache. If the test starts, it means the `Asset.find_asset_file()` utility will succeed when looking for the asset.

[1] Now, if the objective is to make the requirements available to the test without any test interaction (e.g., looking for the requirements locally), I agree we can find a way to build a list of `requirement:target` to pass to the test. Still, again, I don't think it is the _requirements resolver_ responsibility to build each detail of this list; on the other hand, it may be its responsibility to consolidate these pieces of information provided by each requirement module.

> The point of using an environment variable, rather than adding support for the test is based on:
> 
> 1. `exec-test` and similars (`tap`) won't have another way of accessing those requirements

In my point of view, this is a new feature for `exec-test` and `tap` test. Today they work without it, looking for its requirements during the test execution or with a pre-test script. I don't think this is something that impacts this PR, but an extension to it. In this case, we have time to discuss possible solutions.

> 2. current `INSTRUMENTED` tests are being kept compatible, but we may want/need to break compatibility when the current runner is replaced.

Right, and it may use the new feature as I commented in 1, but today, I don't see this PR as an impacting point to it.

Reply:
> From my point of view, the _requirements resolver_ will be responsible for collecting/fulfilling the requirements for a test. This action, as you previously mentioned, will allow a test to start, depending on the availability of all the requirements. Initially, thinking from a high level, I would not expect the _requirements resolver_ to provide any information to the test (the opposite later in this comment [1]).
> 

OK, so what I get from this is that you agree that we should not add a `requirements` parameter and attribute to the `Avocado.Test` class at this point, right?

Reply:
> > From my point of view, the _requirements resolver_ will be responsible for collecting/fulfilling the requirements for a test. This action, as you previously mentioned, will allow a test to start, depending on the availability of all the requirements. Initially, thinking from a high level, I would not expect the _requirements resolver_ to provide any information to the test (the opposite later in this comment [1]).
> 
> OK, so what I get from this is that you agree that we should not add a `requirements` parameter and attribute to the `Avocado.Test` class at this point, right?

Yes, I agree considering this sentence, no, I don't agree considering sentence [1]. :D
It is just a matter of deciding which one we want to follow.

Reply:
OK, before this gets more confusing, let me just say: I suggest you don't export the requirements to the `Avocado.Test` at this time.  Let me know what you decide on doing.

Reply:
> OK, before this gets more confusing, let me just say: I suggest you don't export the requirements to the `Avocado.Test` at this time. Let me know what you decide on doing.

Okay, this (adding the requirements to the test) is a small change we can do later if we decide to go for it.

What about the loader, as you mentioned in the first comment?